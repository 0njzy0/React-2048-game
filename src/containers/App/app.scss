@import '../../assets/styles/vars';

.app {
  margin: 20px;
  text-align: center;
}

.game-over {
  padding-bottom: 1em;
  padding-top: 1em;
}

.box {
  // align-items: flex-end;
  display: flex;
  justify-content: center;
}

.board {
  // flex: $flex;
}

.btn-group {
  align-items: center;
  display: flex;
  flex-direction: column;
  margin-bottom: 2px;
  margin-left: 50px;

  .new-game {
    align-items: center;
    display: flex;
    flex-grow: 1;
  }

  img {
    color: $white;
    width: 23px;
  }

  .buttons {
    display: flex;
    flex-direction: row;
  }
}
