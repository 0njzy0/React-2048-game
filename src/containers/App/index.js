import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import debounce from 'lodash.debounce';
import styles from './app.scss';
import resetSvg from '../../assets/svg/reset.svg';
import Board from '../../components/Board';
import Modal from '../../components/Modal';
import Firework from '../../components/Firework';
import Tips from '../../components/Tips';
import Footer from '../../components/Footer';
import ControlPanel from '../ControlPanel';
import Scores from '../Scores';
import {
  initMatrix,
  reset,
} from '../../reducers/board';
import i18n from '../../utils/i18n';
import comments from '../../utils/gitComments';

document.title = i18n.title;

class App extends Component {
  static propTypes = {
    onInit: PropTypes.func.isRequired,
    onReset: PropTypes.func.isRequired,

    matrix: PropTypes.arrayOf(PropTypes.array).isRequired,
    score: PropTypes.number.isRequired,
    gameOver: PropTypes.bool.isRequired,
  };

  constructor(...args) {
    super(...args);

    // debounce delay in ms
    this.delay = process.env.NODE_ENV === 'production' ? 300 : 1;
  }

  componentWillMount() {
    const board = JSON.parse(localStorage.getItem('board')) || null;
    if (board && !board.gameOver) {
      this.props.onInit(board);
    } else {
      this.props.onInit();
    }
  }

  componentDidMount() {
    comments.renderHeader(document.getElementById('gitmentHeader'));
    comments.renderComments(document.getElementById('gitmentComments'));
    comments.renderEditor(document.getElementById('gitmentEditor'));
  }

  render() {
    const { delay } = this;
    const { matrix, score, gameOver } = this.props;

    return (
      <div className={styles.app}>
        <div className={styles.box}>
          <div className={styles.board}>
            <h1 className={styles.title}>2048</h1>
            <Board matrix={matrix} />
          </div>
          <div className={styles.panel}>
            <div>
              <Scores />
            </div>
            <ControlPanel delay={delay} />
          </div>
        </div>
        <Tips title={i18n.tipTitle} content={i18n.tipContent} />
        <div className={styles.comments} >
          {/* eslint-disable react/no-danger */}
          <p
            className={styles.commentsTitle}
            dangerouslySetInnerHTML={{ __html: i18n.commentTitle }}
          />
          <div id="gitmentHeader" />
          <div id="gitmentComments" />
          <div id="gitmentEditor" />
        </div>
        <Footer
          name={'devrsi0n'}
          profileUrl={'https://github.com/devrsi0n'}
          repoUrl={'https://github.com/devrsi0n/React-2048-game'}
        />
        {/* <Modal display >{gameOver} */}
        <Modal display={gameOver} >
          {
            score > 999 ? <Firework /> : null
          }
          <div>
            <div className={styles.gameover}>
              <div className={styles.text}> {i18n.score} </div>
              <div className={styles.banner}>
                <p>{score}</p>
              </div>
            </div>
            <div className={styles.buttonWrapper}>
              <button
                className={styles.button}
                onClick={debounce(this.props.onReset, delay)}
              >
                <img src={resetSvg} alt="reset" />
              </button>
            </div>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  matrix: state.board.matrix,
  score: state.board.score,
  gameOver: state.board.gameOver,
});
const mapDispatchToProps = dispatch => ({
  onInit(board) {
    dispatch(initMatrix(board));
  },
  onReset() {
    dispatch(reset());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
