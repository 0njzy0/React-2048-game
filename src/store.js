import { createStore } from 'redux';
import undoable from 'redux-undo';
import rootReducer from './reducers';

const initHistory = JSON.parse(localStorage.getItem('state') || 'null');
/* eslint-disable no-underscore-dangle */
const args = [
  undoable(rootReducer, {
    limit: 11, // set a limit for the history
    ignoreInitialState: true,
  }),
  window.__REDUX_DEVTOOLS_EXTENSION__ &&
  window.__REDUX_DEVTOOLS_EXTENSION__(),
];
if (initHistory) {
  args.splice(1, 0, initHistory);
}

const store = createStore(...args);

store.subscribe(() => {
  const state = store.getState();
  localStorage.setItem('state', JSON.stringify(state));
});

export default store;
