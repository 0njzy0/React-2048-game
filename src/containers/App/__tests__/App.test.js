import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import App from '..';
import store from '../../../store';

// Make random always return 0.5, so snapshot always same.
const mockMath = Object.create(global.Math);
mockMath.random = () => 0.5;
global.Math = mockMath;

describe('<App />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Provider store={store} >
        <App />
      </Provider>, div);
  });

  it('component render', () => {
    const app = renderer.create(
      <Provider store={store} >
        <App />
      </Provider>,
    ).toJSON();
    expect(app).toMatchSnapshot();
  });
});

