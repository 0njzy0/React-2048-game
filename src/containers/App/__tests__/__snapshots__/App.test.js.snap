// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobileApp /> component render 1`] = `
<div>
  <div
    className="app"
  >
    <div
      className="box"
    >
      <div
        className="board"
      >
        <h1
          className="title"
        >
          2048
        </h1>
        <table
          className="board"
        >
          <tbody>
            <tr>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
            </tr>
            <tr>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell color-4"
                >
                  <div
                    className="number"
                  >
                    4
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td>
                <div
                  className="cell color-4"
                >
                  <div
                    className="number"
                  >
                    4
                  </div>
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
            </tr>
            <tr>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
              <td>
                <div
                  className="cell"
                >
                  <div
                    className="number"
                  />
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="panel"
      >
        <div>
          <div
            className="index"
          >
            <p
              className="score"
            >
              SCORE
              <em
                className="text"
              >
                0
              </em>
            </p>
            <p
              className="score best"
            >
              BEST
              <em
                className="text"
              >
                0
              </em>
            </p>
          </div>
        </div>
        <div
          className="panel"
        >
          <div
            className="rowBtn"
          >
            <div
              className="btn"
            >
              <button
                className="btn-flat btn-md btn-default"
                onClick={[Function]}
              >
                <img
                  alt="speaker"
                  src="speaker-on.svg"
                />
              </button>
            </div>
            <div
              className="btn "
            >
              <button
                className="btn-flat btn-md btn-default"
                onClick={[Function]}
              >
                <img
                  alt="undo"
                  src="undo.svg"
                />
              </button>
            </div>
            <div
              className="btn "
            >
              <button
                className="btn-flat btn-md btn-default"
                onClick={[Function]}
              >
                <img
                  alt="reset"
                  src="reset.svg"
                />
              </button>
            </div>
          </div>
          <div
            className="btn "
          >
            <button
              className="btn-flat btn-md btn-primary"
              onClick={[Function]}
            >
              <img
                alt="arrow up"
                className="up"
                src="arrow.svg"
              />
            </button>
          </div>
          <div
            className="arrows"
          >
            <div
              className="btn "
            >
              <button
                className="btn-flat btn-md btn-primary"
                onClick={[Function]}
              >
                <img
                  alt="arrow left"
                  className="left"
                  src="arrow.svg"
                />
              </button>
            </div>
            <div
              className="btn "
            >
              <button
                className="btn-flat btn-md btn-primary"
                onClick={[Function]}
              >
                <img
                  alt="arrow down"
                  className="down"
                  src="arrow.svg"
                />
              </button>
            </div>
            <div
              className="btn "
            >
              <button
                className="btn-flat btn-md btn-primary"
                onClick={[Function]}
              >
                <img
                  alt="arrow right"
                  className="right"
                  src="arrow.svg"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="tips"
    >
      <p
        className="title"
      >
        Tips
      </p>
      <p
        className="content"
      >
        Use keyboard arrow keys (or \`WASD\` if you like) control blocks;Click undo button to revert to last step status if you regrets
      </p>
    </div>
    <div
      className="footer"
    >
      <div
        className="footer"
      >
        <a
          className="icon"
          href="https://github.com/devrsi0n/React-2048-game"
        >
          <img
            alt="github repo"
            className="github"
            src="github.png"
          />
        </a>
        Build with
        <span
          className="heart"
        >
          â™¥
        </span>
        by
        <a
          className="link"
          href="https://github.com/devrsi0n"
        >
          devrsi0n
        </a>
      </div>
    </div>
    <div>
      <div
        className="hidden"
      >
        <div
          className="overlay"
        />
        <div
          className="modal center"
        >
          <div>
            <div
              className="gameover"
            >
              <div
                className="text"
              >
                 
                SCORE
                 
              </div>
              <div
                className="banner"
              >
                <p>
                  0
                </p>
              </div>
            </div>
            <div
              className="buttonWrapper"
            >
              <button
                className="button"
                onClick={[Function]}
              >
                <img
                  alt="reset"
                  src="reset.svg"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
